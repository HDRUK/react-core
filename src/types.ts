export type CommonVariants = "primary" | "secondary" | "tertiary";

export type CommonSizes = "small" | "medium" | "large";

export interface CommonStyleProps {
  display?: string;
  alignItems?: string;
  justifyContent?: string;
  p?: number;
  pl?: number;
  pb?: number;
  pt?: number;
  pr?: number;
  m?: number;
  ml?: number;
  mb?: number;
  mt?: number;
  mr?: number;
  flexGrow?: string;
  position?: string;
  width?: string;
  minWidth?: string;
  maxWidth?: string;
  height?: string;
  minHeight?: string;
  maxHeight?: string;
}

export type ComponentStyleProps = MarginStyleProps & WidthStyleProps;

export interface MarginStyleProps {
  m?: number;
  ml?: number;
  mb?: number;
  mt?: number;
  mr?: number;
}

export interface WidthStyleProps {
  width?: string;
  minWidth?: string;
  maxWidth?: string;
}

export interface HeightStyleProps {
  height?: string;
  minHeight?: string;
  maxHeight?: string;
}

export interface ThemeFontSizes {
  xxs: string;
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
  "2xl": string;
  "3xl": string;
  "4xl": string;
  "5xl": string;
}

export interface ThemeColors {
  [key: string]: string;
}

export interface ThemeComponents {
  [key: string]: string;
}

export interface Theme {
  base: {
    increment: number;
    unit: string;
  };
  font: {
    size: ThemeFontSizes;
  };
  colors: ThemeColors;
  components: ThemeComponents;
}
