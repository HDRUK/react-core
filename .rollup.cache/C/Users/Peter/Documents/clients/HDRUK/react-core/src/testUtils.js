// @ts-nocheck
import { __assign } from "tslib";
import React from "react";
import { ThemeProvider } from "@emotion/react";
import { render } from "@testing-library/react";
import { DEFAULT_THEME } from "./theme";
export var Providers = function (_a) {
    var children = _a.children;
    return React.createElement(ThemeProvider, { theme: DEFAULT_THEME }, children);
};
export var renderWithProviders = function (ui, options) {
    return render(ui, __assign({ wrapper: Providers }, options));
};
//# sourceMappingURL=testUtils.js.map