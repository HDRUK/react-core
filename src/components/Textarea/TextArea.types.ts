import { ReactNode, RefObject } from "react";
import { TextareaAutosizeProps } from "react-textarea-autosize";
import { ComponentHeightProps, ComponentStyleProps } from "../../types";
import { InputVariants } from "../Input/Input.types";

export interface TextAreaProps
    extends ComponentStyleProps,
        ComponentHeightProps,
        Omit<TextareaAutosizeProps, "width" | "height" | "value"> {
    autosize: boolean;
    maxCharCount: number;
    charCountDescription: ReactNode;
    variant: InputVariants;
    description?: ReactNode;
    error?: ReactNode;
    label?: ReactNode;
    inputRef?: RefObject<HTMLTextAreaElement>;
    value?: string;
}

export type TextAreaStyleProps = Pick<
    TextAreaProps,
    "variant" | "minHeight" | "maxHeight" | "error"
>;
