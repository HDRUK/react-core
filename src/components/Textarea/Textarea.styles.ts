import { css, Theme } from "@emotion/react";
import { input } from "../Input/Input.styles";
import { TextAreaStyleProps } from "./TextArea.types";

export const root =
    ({ variant, minHeight, maxHeight, error }: TextAreaStyleProps) =>
    (theme: Theme) => {
        return css`
            width: 100%;
            flex-direction: column;
            flex: 0 0 100%;

            textarea {
                ${input({ variant, error })(theme)}

                width: 100%;
                min-height: ${minHeight};
                max-height: ${maxHeight};
            }
        `;
    };
