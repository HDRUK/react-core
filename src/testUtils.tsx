import React from "react";
import { ThemeProvider } from "@emotion/react";
import { render, RenderOptions } from "@testing-library/react";
import { DEFAULT_THEME } from "./theme";

export const Providers: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  return <ThemeProvider theme={DEFAULT_THEME}>{children}</ThemeProvider>;
};

export const renderWithProviders = (
  ui: React.ReactElement,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};
