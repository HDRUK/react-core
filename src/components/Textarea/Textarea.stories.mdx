import { Meta, Story, Canvas } from "@storybook/addon-docs";
import { useState } from "react";

import Textarea from "./Textarea";

import Box from "../Box";
import { Card, CardBody } from "../Card";

<Meta title="Molecules/Textarea" component={Textarea} />

<Canvas>
    <Story name="Basic">
        {props => (
            <Card>
                <CardBody>
                    <Box width="300px">
                        <Textarea label="Primary" mb={4} />
                        <Textarea label="Primary (disabled)" disabled mb={4} />
                        <Textarea
                            label="Primary with error"
                            error="Must have a value"
                            mb={10}
                        />
                        <Textarea
                            variant="secondary"
                            label="Secondary"
                            mb={4}
                        />
                        <Textarea
                            variant="secondary"
                            label="Secondary (disabled)"
                            mb={4}
                            disabled
                        />
                        <Textarea
                            variant="secondary"
                            label="Basic with error"
                            error="Must have a value"
                        />
                    </Box>
                </CardBody>
            </Card>
        )}
    </Story>
    <Story name="With count">
        {props => {
            const [value, setValue] = useState();
            const handleChange = ({ target: { value } }) => {
                setValue(value);
            };
            return (
                <Card>
                    <CardBody>
                        <Box width="300px">
                            <Textarea
                                label="Count label"
                                maxCharCount={255}
                                charCountDescription="characters of (example)"
                                value={value}
                                onChange={handleChange}
                            />
                        </Box>
                    </CardBody>
                </Card>
            );
        }}
    </Story>
    <Story name="Autosize">
        {props => {
            const [value, setValue] = useState();
            const handleChange = ({ target: { value } }) => {
                setValue(value);
            };
            return (
                <Card>
                    <CardBody>
                        <Box width="300px">
                            <Textarea
                                label="Autosize"
                                autosize
                                height="100px"
                                value={value}
                                onChange={handleChange}
                            />
                        </Box>
                    </CardBody>
                </Card>
            );
        }}
    </Story>
</Canvas>
