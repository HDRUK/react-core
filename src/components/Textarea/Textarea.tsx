/** @jsxImportSource @emotion/react */
import { cx } from "@emotion/css";
import TextareaAutosize from "react-textarea-autosize";
import useCommonStyles from "../../hooks/useCommonStyles";
import Box from "../Box";
import Label from "../Label";
import Message from "../Message";
import Typography from "../Typography";
import * as styles from "./Textarea.styles";
import { TextAreaProps } from "./TextArea.types";

const Textarea = ({
    className,
    autosize,
    label,
    maxCharCount,
    charCountDescription,
    variant,
    value,
    mt,
    mb,
    ml,
    mr,
    m,
    mx,
    my,
    width,
    minWidth,
    maxWidth,
    minHeight,
    maxHeight,
    height,
    inputRef,
    id,
    error,
    disabled,
    ...outerProps
}: TextAreaProps) => {
    const commonStyles = useCommonStyles({
        mt,
        mb,
        ml,
        mr,
        m,
        mx,
        my,
        width,
        minWidth,
        maxWidth,
        height,
    });

    return (
        <div
            className={cx("ui-Textarea", className)}
            css={[
                commonStyles,
                styles.root({ variant, minHeight, maxHeight, error }),
            ]}>
            {label && (
                <Label htmlFor={id} mb={2} disabled={disabled}>
                    {label}
                </Label>
            )}
            <div>
                {!!maxCharCount && (
                    <Typography color="grey700Alt">
                        <Box mb={1} display="flex">
                            {value?.length || 0} {charCountDescription}
                            <Box textAlign="right" flexGrow="1">
                                ({value?.length || 0}/{maxCharCount})
                            </Box>
                        </Box>
                    </Typography>
                )}
                {autosize && (
                    <TextareaAutosize
                        value={value}
                        id={id}
                        disabled={disabled}
                        ref={inputRef}
                        maxLength={maxCharCount}
                        {...outerProps}
                    />
                )}
                {!autosize && (
                    <textarea
                        id={id}
                        disabled={disabled}
                        ref={inputRef}
                        maxLength={maxCharCount}
                        {...outerProps}
                    />
                )}
            </div>
            {error && (
                <Message variant="danger" mt={2}>
                    {error}
                </Message>
            )}
        </div>
    );
};

Textarea.defaultProps = {
    autosize: false,
    variant: "primary",
    charCountDescription: "character limit",
    value: "",
    minHeight: "80px",
};

export default Textarea;
