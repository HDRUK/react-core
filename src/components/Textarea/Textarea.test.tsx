import { fireEvent, RenderResult } from "@testing-library/react";
import React from "react";
import Textarea from ".";
import { renderWithProviders } from "../../testUtils";
import { TextAreaProps } from "./TextArea.types";

const props: TextAreaProps = {
    variant: "primary",
    label: "Sample label",
    charCountDescription: "sample description",
    onChange: jest.fn(),
};

let wrapper: RenderResult;

describe("Given the Textarea component", () => {
    describe("When it is rendered", () => {
        beforeEach(() => {
            wrapper = renderWithProviders(<Textarea {...props} />);
        });

        it("Then should match the snapshot", () => {
            expect(wrapper.asFragment()).toMatchSnapshot();
        });

        it("Then has the correct label", async () => {
            expect(wrapper.getByText("Sample label")).toBeTruthy();
        });

        describe("And it is countable", () => {
            beforeEach(() => {
                wrapper.rerender(<Textarea {...props} maxCharCount={5} />);
            });

            it("Then should match the snapshot", () => {
                expect(wrapper.asFragment()).toMatchSnapshot();
            });

            it("Then has the correct count text", async () => {
                expect(wrapper.getByTestId("count")).toHaveTextContent(
                    "0 sample description (0/5)"
                );
            });

            describe("And the value is changed", () => {
                beforeEach(() => {
                    wrapper.rerender(
                        <Textarea {...props} maxCharCount={5} value="1234" />
                    );

                    const close = wrapper.getByRole("textbox");

                    fireEvent.change(close, { target: { value: "value" } });
                });

                it("Then calls onChange", async () => {
                    expect(props.onChange).toHaveBeenCalled();
                });

                it("Then has the correct count text", async () => {
                    expect(wrapper.getByTestId("count")).toHaveTextContent(
                        "4 sample description (4/5)"
                    );
                });

                describe("And it is too long", () => {
                    beforeEach(() => {
                        wrapper.rerender(
                            <Textarea
                                {...props}
                                maxCharCount={5}
                                value="123456"
                            />
                        );
                    });

                    it("Then has the correct count text", async () => {
                        expect(wrapper.getByTestId("count")).toHaveTextContent(
                            "5 sample description (5/5)"
                        );
                    });
                });
            });
        });

        describe("And it has autosize", () => {
            beforeEach(() => {
                wrapper.rerender(<Textarea {...props} autosize />);
            });

            it("Then should match the snapshot", () => {
                expect(wrapper.asFragment()).toMatchSnapshot();
            });
        });
    });
});
