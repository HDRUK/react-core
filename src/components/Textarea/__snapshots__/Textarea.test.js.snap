// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Textarea component When it is rendered Then matches the previous snapshot 1`] = `
<div>
  <div
    class="ui-Box css-ichth-B"
  >
    <div
      class="ui-Textarea additional-classname css-1r4a1n6-Textarea form-group"
    >
      <label
        class="css-8ro76h-Textarea form-label"
      >
        Sample label
      </label>
      <div
        class="css-v8dvh9-Textarea input-group"
      >
        <div
          class="ui-TextArea__charCount css-25vzsz-Textarea"
        >
          11
           
          sample description
          <span
            class="css-ghohkr-Textarea"
          >
             
            (
            11
            /
            255
            )
          </span>
        </div>
        <textarea
          class="form-control"
        />
      </div>
    </div>
  </div>
</div>
`;
